tensorstore/kvstore/s3/s3_key_value_store.cc

```diff
@@ -0,0 +1,1461 @@
+    future.ExecuteWhenReady(WithExecutor(
+        owner_->executor(), [self = IntrusivePtr<ListTask>(this)](
+                                ReadyFuture<HttpResponse> response) {
+          self->OnResponse(response.result());
+        }));
+  }
+
+  void OnResponse(const Result<HttpResponse>& response) {
+    auto status = OnResponseImpl(response);
+    // OkStatus are handled by OnResponseImpl
+    if (absl::IsCancelled(status)) {
+      execution::set_done(receiver_);
+      execution::set_stopping(receiver_);
+      return;
+    }
+    if (!status.ok()) {
+      execution::set_error(receiver_, std::move(status));
+      execution::set_stopping(receiver_);
+      return;
+    }
+  }
+
+  /// @brief Find the starting position of the tag or the position immediately after the tag
+  /// with the supplied XML payload
+  /// @param data XML payload
+  /// @param tag XML tag
+  /// @param pos Initial searching position
+  /// @param start true if starting position desired otherwise false
+  /// @return The tag starting position, otherwise the position immediately after the tag
+  Result<std::size_t> FindTag(std::string_view data, std::string_view tag,
+                              std::size_t pos=0, bool start=true) {
+    if(pos=data.find(tag, pos); pos != std::string_view::npos) {
+      return start ? pos :  pos + tag.length();
+    }
+    return absl::NotFoundError(
+      absl::StrCat("Malformed List Response XML: can't find ", tag, " in ", data));
+  }
+
+  /// @brief Get tag contents within the supplied XML payload
+  /// @param data XML payload
+  /// @param open_tag Opening tag
+  /// @param close_tag Closing tag
+  /// @param pos Initial searching position. Updated with position immediately
+  ///            after the closing tag.
+  /// @return Tag contents
+  Result<std::string_view> GetTag(std::string_view data,
+                                  std::string_view open_tag,
+                                  std::string_view close_tag,
+                                  std::size_t * pos) {
+
+    TENSORSTORE_ASSIGN_OR_RETURN(auto tagstart, FindTag(data, open_tag, *pos, false));
+    TENSORSTORE_ASSIGN_OR_RETURN(auto tagend, FindTag(data, close_tag, tagstart, true));
+    *pos = tagend + close_tag.size();
+    return data.substr(tagstart, tagend - tagstart);
+  }
+
+
+  absl::Status OnResponseImpl(const Result<HttpResponse>& response) {
+    if (is_cancelled()) {
+      return absl::CancelledError();
+    }
+    ABSL_LOG_IF(INFO, TENSORSTORE_INTERNAL_S3_LOG_RESPONSES && response.ok())
+        << "List " << *response;
+
+    absl::Status status =
+        response.ok() ? HttpResponseCodeToStatus(*response) : response.status();
+    if (!status.ok() && IsRetriable(status)) {
+      return owner_->BackoffForAttemptAsync(status, attempt_++, this);
+    }
+
+    auto cord = response->payload;
+    auto payload = cord.Flatten();
+    auto kListBucketOpenTag = "<ListBucketResult xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">";
```

Comment:
I went with simple string search here mostly because RE2 doesn't support backreferences. Otherwise, something like https://flapenguin.me/xml-regex would have been appealing.

The string search does restart searches for outer level tags from a starting position but there aren't many of these and the actual [S3 XML Response](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_ResponseSyntax) seems to be well-behaved in that that the sequence of `<Content>` tags and associated inner tags (`<Key>`, `<Etag>`) are placed after the outer tags. Searches for the next `<Content>` tag (and `<Key>`) do start from a previous seach position.

Reply:
Your approach seems reasonable.